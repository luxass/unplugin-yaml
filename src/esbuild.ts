/**
 * This entry file is for esbuild plugin. Requires esbuild >= 0.15
 *
 * @module
 */

import { createEsbuildPlugin } from "unplugin";
import type unplugin from "./";
import { unpluginFactory } from "./";

/**
 * Esbuild plugin
 *
 * @example
 * ```ts
 * // esbuild.config.js
 * import { build } from "esbuild";
 *
 * build({
 *   plugins: [require("unplugin-yaml/esbuild")()],
 * })
 * ```
 */
export default createEsbuildPlugin(unpluginFactory) as typeof unplugin.esbuild;
