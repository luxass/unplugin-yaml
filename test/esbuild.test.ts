import { join } from "node:path";
import { dedent } from "@luxass/utils";
import { build } from "esbuild";
import { describe, expect, it } from "vitest";
import { testdir } from "vitest-testdirs";
import YAMLPlugin from "../src/esbuild";

describe("esbuild", () => {
  it("expect yaml imports to be json objects", async () => {
    const testdirPath = await testdir.from(join(import.meta.dirname, "fixtures/basic"));

    expect(testdirPath).toBeDefined();

    await build({
      entryPoints: [
        join(testdirPath, "basic.js"),
      ],
      format: "esm",
      outfile: join(testdirPath, "output.js"),
      bundle: true,
      minifySyntax: false,
      plugins: [
        YAMLPlugin(),
      ],
    });

    const module = await import(join(testdirPath, "output.js")).then((m) => m);
    expect(module).toBeDefined();

    const matchedCfg = {
      pluginDir: "./plugins",
      web: { enabled: true },
      logging: { type: "stdout", level: "info" },
    };

    expect(module.yaml).toEqual(matchedCfg);
    expect(module.yml).toEqual(matchedCfg);
  });

  it("expect yaml imports to be strings", async () => {
    const testdirPath = await testdir.from(join(import.meta.dirname, "fixtures/basic"));

    expect(testdirPath).toBeDefined();

    await build({
      entryPoints: [
        join(testdirPath, "basic-raw.js"),
      ],
      format: "esm",
      outfile: join(testdirPath, "output-raw.js"),
      bundle: true,
      minifySyntax: false,
      plugins: [
        YAMLPlugin(),
      ],
    });

    const module = await import(join(testdirPath, "output-raw.js")).then((m) => m);
    expect(module).toBeDefined();

    const expectedString = dedent`
      pluginDir: ./plugins

      web:
        enabled: true

      logging:
        type: stdout
        level: info
    `;

    expect(module.yml).toMatch(expectedString);
    expect(module.yaml).toMatch(expectedString);
  });

  it("handle transforms", async () => {
    const testdirPath = await testdir.from(join(import.meta.dirname, "fixtures/transform"));

    expect(testdirPath).toBeDefined();

    await build({
      entryPoints: [
        join(testdirPath, "transform.js"),
      ],
      format: "esm",
      outfile: join(testdirPath, "output-transform.js"),
      bundle: true,
      minifySyntax: false,
      plugins: [
        YAMLPlugin({
          transform(data) {
            if (data != null && typeof data === "object" && "this" in data) {
              return {
                this: "transformed",
              };
            }
          },
        }),
      ],
    });

    const config = await import(join(testdirPath, "output-transform.js")).then((m) => m.config);
    expect(config).toBeDefined();

    expect(config).toEqual({
      this: "transformed",
    });
  });

  it("handle multi document", async () => {
    const testdirPath = await testdir.from(join(import.meta.dirname, "fixtures/multi"));

    expect(testdirPath).toBeDefined();

    await build({
      entryPoints: [
        join(testdirPath, "multi.js"),
      ],
      format: "esm",
      outfile: join(testdirPath, "output-multi.js"),
      bundle: true,
      minifySyntax: false,
      plugins: [
        YAMLPlugin({
          type: "multi",
        }),
      ],
    });

    const cronjobs = await import(join(testdirPath, "output-multi.js")).then((m) => m.cronjobs);
    expect(cronjobs).toBeDefined();

    expect(cronjobs).toEqual([
      {
        apiVersion: "batch/v1",
        kind: "CronJob",
        metadata: {
          name: "first-cronjob",
          namespace: "default",
        },
        spec: {
          concurrencyPolicy: "Forbid",
          failedJobsHistoryLimit: 1,
          jobTemplate: {
            spec: {
              template: {
                spec: {
                  containers: [
                    {
                      command: [
                        "/bin/sh",
                        "-c",
                        "echo \"First job running\"; date",
                      ],
                      image: "busybox:latest",
                      name: "first-job",
                    },
                  ],
                  restartPolicy: "OnFailure",
                },
              },
            },
          },
          schedule: "*/10 * * * *",
          successfulJobsHistoryLimit: 3,
        },
      },
      {
        apiVersion: "batch/v1",
        kind: "CronJob",
        metadata: {
          name: "second-cronjob",
          namespace: "default",
        },
        spec: {
          concurrencyPolicy: "Forbid",
          failedJobsHistoryLimit: 1,
          jobTemplate: {
            spec: {
              template: {
                spec: {
                  containers: [
                    {
                      command: [
                        "/bin/sh",
                        "-c",
                        "echo \"Second job running\"; date",
                      ],
                      image: "busybox:latest",
                      name: "second-job",
                    },
                  ],
                  restartPolicy: "OnFailure",
                },
              },
            },
          },
          schedule: "0 0 * * *",
          successfulJobsHistoryLimit: 3,
        },
      },
    ]);
  });
});
