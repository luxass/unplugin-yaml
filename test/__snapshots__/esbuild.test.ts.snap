// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`handle multi document 1`] = `
"var data = [
  {
    "apiVersion": "batch/v1",
    "kind": "CronJob",
    "metadata": {
      "name": "first-cronjob",
      "namespace": "default"
    },
    "spec": {
      "schedule": "*/10 * * * *",
      "concurrencyPolicy": "Forbid",
      "successfulJobsHistoryLimit": 3,
      "failedJobsHistoryLimit": 1,
      "jobTemplate": {
        "spec": {
          "template": {
            "spec": {
              "containers": [
                {
                  "name": "first-job",
                  "image": "busybox:latest",
                  "command": [
                    "/bin/sh",
                    "-c",
                    'echo "First job running"; date'
                  ]
                }
              ],
              "restartPolicy": "OnFailure"
            }
          }
        }
      }
    }
  },
  {
    "apiVersion": "batch/v1",
    "kind": "CronJob",
    "metadata": {
      "name": "second-cronjob",
      "namespace": "default"
    },
    "spec": {
      "schedule": "0 0 * * *",
      "concurrencyPolicy": "Forbid",
      "successfulJobsHistoryLimit": 3,
      "failedJobsHistoryLimit": 1,
      "jobTemplate": {
        "spec": {
          "template": {
            "spec": {
              "containers": [
                {
                  "name": "second-job",
                  "image": "busybox:latest",
                  "command": [
                    "/bin/sh",
                    "-c",
                    'echo "Second job running"; date'
                  ]
                }
              ],
              "restartPolicy": "OnFailure"
            }
          }
        }
      }
    }
  }
];
var cronjobs_default = data;

console.log(cronjobs_default);
"
`;

exports[`handle transforms 1`] = `
"var data = {
  "this": "transformed"
};
var config_default = data;

console.log(config_default);
"
`;

exports[`handles yaml > expect yaml import to be a json object 1`] = `
"var data = {
  "pluginDir": "./plugins",
  "web": {
    "enabled": true
  },
  "logging": {
    "type": "stdout",
    "level": "info"
  }
};
var yaml_config_default = data;

console.log(yaml_config_default);
"
`;

exports[`handles yaml > expect yaml import to be a string 1`] = `
"var yaml_config_default = "pluginDir: ./plugins\\n\\nweb:\\n  enabled: true\\n\\nlogging:\\n  type: stdout\\n  level: info\\n";

console.log(yaml_config_default);
"
`;

exports[`handles yml > expect yml import to be a json object 1`] = `
"var data = {
  "pluginDir": "./plugins",
  "web": {
    "enabled": true
  },
  "logging": {
    "type": "stdout",
    "level": "info"
  }
};
var yml_config_default = data;

console.log(yml_config_default);
"
`;

exports[`handles yml > expect yml import to be a string 1`] = `
"var yml_config_default = "pluginDir: ./plugins\\n\\nweb:\\n  enabled: true\\n\\nlogging:\\n  type: stdout\\n  level: info\\n";

console.log(yml_config_default);
"
`;
